File: 138. Copy List with Random Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        Node* headOfCopy = NULL, *prv = NULL;
        unordered_map <Node*, Node*> ump;
        Node* cur = head;
        while( cur != NULL ) { 
            Node* curNode = new Node(cur->val);
            
            if( prv == NULL ) { 
                headOfCopy = curNode;
            } else { 
                prv->next = curNode;
            }
            prv = curNode;
            ump[cur] = curNode;
            cur = cur->next;
        }
        cur = head;
        Node* copyCur = headOfCopy;
        while( cur != NULL ) { 
            copyCur->random = ump[cur->random];
            cur = cur->next;
            copyCur = copyCur->next;
        }
        return headOfCopy;
    }
};

File: Java/138. Copy List with Random Pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        HashMap<Node, Node> map = new HashMap<>();
        Node newHead = null, prv = null;
        Node curNode = head;
        while( curNode != null ) {
            Node newNode = new Node(curNode.val);
            if( newHead == null ) {
                newHead = newNode;
            } else {
                prv.next = newNode;
            }
            map.put(curNode, newNode);
            curNode = curNode.next;
            prv = newNode;
        }
        curNode = head;
        Node newCurNode = newHead;
        while( curNode != null ) {
            newCurNode.random = map.get(curNode.random);
            curNode = curNode.next;
            newCurNode = newCurNode.next;
        }
        return newHead;
    }
}
